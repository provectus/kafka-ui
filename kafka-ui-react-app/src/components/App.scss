$header-height: 52px;
$navbar-width: 250px;

.Layout {
  &__header {
    box-shadow: 0 0.46875rem 2.1875rem rgba(4,9,20,0.03),
      0 0.9375rem 1.40625rem rgba(4,9,20,0.03),
      0 0.25rem 0.53125rem rgba(4,9,20,0.05),
      0 0.125rem 0.1875rem rgba(4,9,20,0.03);
    z-index: 31;
  }

  &__container {
    margin-top: $header-height;
    margin-left: $navbar-width;
  }

  &__navbar {
    width: $navbar-width;
    display: flex;
    flex-direction: column;
    box-shadow: 7px 0 60px rgba(0,0,0,0.05);
    position: fixed;
    top: $header-height;
    left: 0;
    bottom: 0;
    padding: 20px 20px;
    overflow-y: scroll;
  }
}
