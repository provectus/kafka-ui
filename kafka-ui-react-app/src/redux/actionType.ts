enum ActionType {
  GET_CLUSTERS__REQUEST = 'GET_CLUSTERS__REQUEST',
  GET_CLUSTERS__SUCCESS = 'GET_CLUSTERS__SUCCESS',
  GET_CLUSTERS__FAILURE = 'GET_CLUSTERS__FAILURE',

  GET_BROKERS__REQUEST = 'GET_BROKERS__REQUEST',
  GET_BROKERS__SUCCESS = 'GET_BROKERS__SUCCESS',
  GET_BROKERS__FAILURE = 'GET_BROKERS__FAILURE',

  GET_BROKER_METRICS__REQUEST = 'GET_BROKER_METRICS__REQUEST',
  GET_BROKER_METRICS__SUCCESS = 'GET_BROKER_METRICS__SUCCESS',
  GET_BROKER_METRICS__FAILURE = 'GET_BROKER_METRICS__FAILURE',

  GET_TOPICS__REQUEST = 'GET_TOPICS__REQUEST',
  GET_TOPICS__SUCCESS = 'GET_TOPICS__SUCCESS',
  GET_TOPICS__FAILURE = 'GET_TOPICS__FAILURE',

  GET_TOPIC_MESSAGES__REQUEST = 'GET_TOPIC_MESSAGES__REQUEST',
  GET_TOPIC_MESSAGES__SUCCESS = 'GET_TOPIC_MESSAGES__SUCCESS',
  GET_TOPIC_MESSAGES__FAILURE = 'GET_TOPIC_MESSAGES__FAILURE',

  GET_TOPIC_DETAILS__REQUEST = 'GET_TOPIC_DETAILS__REQUEST',
  GET_TOPIC_DETAILS__SUCCESS = 'GET_TOPIC_DETAILS__SUCCESS',
  GET_TOPIC_DETAILS__FAILURE = 'GET_TOPIC_DETAILS__FAILURE',

  GET_TOPIC_CONFIG__REQUEST = 'GET_TOPIC_CONFIG__REQUEST',
  GET_TOPIC_CONFIG__SUCCESS = 'GET_TOPIC_CONFIG__SUCCESS',
  GET_TOPIC_CONFIG__FAILURE = 'GET_TOPIC_CONFIG__FAILURE',

  POST_TOPIC__REQUEST = 'POST_TOPIC__REQUEST',
  POST_TOPIC__SUCCESS = 'POST_TOPIC__SUCCESS',
  POST_TOPIC__FAILURE = 'POST_TOPIC__FAILURE',

  PATCH_TOPIC__REQUEST = 'PATCH_TOPIC__REQUEST',
  PATCH_TOPIC__SUCCESS = 'PATCH_TOPIC__SUCCESS',
  PATCH_TOPIC__FAILURE = 'PATCH_TOPIC__FAILURE',

  GET_CONSUMER_GROUPS__REQUEST = 'GET_CONSUMER_GROUPS__REQUEST',
  GET_CONSUMER_GROUPS__SUCCESS = 'GET_CONSUMER_GROUPS__SUCCESS',
  GET_CONSUMER_GROUPS__FAILURE = 'GET_CONSUMER_GROUPS__FAILURE',

  GET_CONSUMER_GROUP_DETAILS__REQUEST = 'GET_CONSUMER_GROUP_DETAILS__REQUEST',
  GET_CONSUMER_GROUP_DETAILS__SUCCESS = 'GET_CONSUMER_GROUP_DETAILS__SUCCESS',
  GET_CONSUMER_GROUP_DETAILS__FAILURE = 'GET_CONSUMER_GROUP_DETAILS__FAILURE',
}

export default ActionType;
