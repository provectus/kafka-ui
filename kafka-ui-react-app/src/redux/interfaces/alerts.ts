import { ErrorResponse } from 'generated-sources';
import React from 'react';

export interface ServerResponse {
  status: number;
  statusText: string;
  body: ErrorResponse;
}

export interface FailurePayload {
  title: string;
  message?: string;
  subject: string;
  subjectId?: string | number;
  response?: ServerResponse;
}

export interface Alert {
  id: string;
  type: 'error' | 'success' | 'warning' | 'info';
  title: string;
  message: React.ReactNode;
  response?: ServerResponse;
  createdAt: number;
}

export type Alerts = Alert[];

export type AlertsState = Record<Alert['id'], Alert>;
