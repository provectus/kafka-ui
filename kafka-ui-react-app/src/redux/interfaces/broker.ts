import { Broker } from 'generated-sources';

export type BrokerId = string;

// export interface Broker {
//   brokerId: BrokerId;
//   bytesInPerSec: number;
//   segmentSize: number;
//   partitionReplicas: number;
//   bytesOutPerSec: number;
// };

export enum ZooKeeperStatus { offline, online };

export interface BrokerDiskUsage {
  brokerId: BrokerId;
  segmentSize: number;
}

export interface BrokerMetrics {
  brokerCount: number;
  zooKeeperStatus: ZooKeeperStatus;
  activeControllers: number;
  onlinePartitionCount: number;
  offlinePartitionCount: number;
  inSyncReplicasCount: number,
  outOfSyncReplicasCount: number,
  underReplicatedPartitionCount: number;
  diskUsage: BrokerDiskUsage[];
}

export interface BrokersState extends BrokerMetrics {
  items: Broker[];
}
