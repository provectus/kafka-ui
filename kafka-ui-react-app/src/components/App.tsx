import React, { Suspense } from 'react';
import { Routes, Route, Navigate } from 'react-router-dom';
import {
  accessErrorPage,
  clusterPath,
  errorPage,
  getNonExactPath,
  clusterNewConfigPath,
} from 'lib/paths';
import PageLoader from 'components/common/PageLoader/PageLoader';
import Dashboard from 'components/Dashboard/Dashboard';
import ClusterPage from 'components/ClusterPage/ClusterPage';
import { ThemeProvider } from 'styled-components';
import { theme, darkTheme } from 'theme/theme';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { showServerError } from 'lib/errorHandling';
import { Toaster } from 'react-hot-toast';
import GlobalCSS from 'components/globalCss';
import * as S from 'components/App.styled';
import ClusterConfigForm from 'widgets/ClusterConfigForm';

import ConfirmationModal from './common/ConfirmationModal/ConfirmationModal';
import { ConfirmContextProvider } from './contexts/ConfirmContext';
import { GlobalSettingsProvider } from './contexts/GlobalSettingsContext';
import ErrorPage from './ErrorPage/ErrorPage';
import { UserInfoRolesAccessProvider } from './contexts/UserInfoRolesAccessContext';
import PageContainer from './PageContainer/PageContainer';

const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      suspense: true,
      onError(error) {
        showServerError(error as Response);
      },
    },
    mutations: {
      onError(error) {
        showServerError(error as Response);
      },
    },
  },
});
const App: React.FC = () => {
  const [isDarkMode, setDarkMode] = React.useState<boolean>(false);

  return (
    <QueryClientProvider client={queryClient}>
      <GlobalSettingsProvider>
        <ThemeProvider theme={isDarkMode ? darkTheme : theme}>
          <Suspense fallback={<PageLoader />}>
            <UserInfoRolesAccessProvider>
              <ConfirmContextProvider>
                <GlobalCSS />
                <S.Layout>
                  <PageContainer setDarkMode={setDarkMode}>
                    <Routes>
                      {['/', '/ui', '/ui/clusters'].map((path) => (
                        <Route
                          key="Home" // optional: avoid full re-renders on route changes
                          path={path}
                          element={<Dashboard />}
                        />
                      ))}
                      <Route
                        path={getNonExactPath(clusterNewConfigPath)}
                        element={<ClusterConfigForm />}
                      />
                      <Route
                        path={getNonExactPath(clusterPath())}
                        element={<ClusterPage />}
                      />
                      <Route
                        path={accessErrorPage}
                        element={
                          <ErrorPage status={403} text="Access is Denied" />
                        }
                      />
                      <Route path={errorPage} element={<ErrorPage />} />
                      <Route
                        path="*"
                        element={<Navigate to={errorPage} replace />}
                      />
                    </Routes>
                  </PageContainer>
                  <Toaster position="bottom-right" />
                </S.Layout>
                <ConfirmationModal />
              </ConfirmContextProvider>
            </UserInfoRolesAccessProvider>
          </Suspense>
        </ThemeProvider>
      </GlobalSettingsProvider>
    </QueryClientProvider>
  );
};

export default App;
