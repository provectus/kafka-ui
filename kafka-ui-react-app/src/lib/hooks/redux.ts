import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux';
import { AppDispatch, RootState } from 'redux/interfaces';

export const useAppDispatch = () => useDispatch<AppDispatch>();
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;
