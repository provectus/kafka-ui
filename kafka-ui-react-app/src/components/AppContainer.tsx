import { connect } from 'react-redux';
import { fetchClustersList } from 'redux/actions';
import {
  getClusterList,
  getIsClusterListFetched,
} from 'redux/reducers/clusters/selectors';
import { getAlerts } from 'redux/reducers/alerts/selectors';
import { RootState } from 'redux/interfaces';
import App from './App';

const mapStateToProps = (state: RootState) => ({
  isClusterListFetched: getIsClusterListFetched(state),
  alerts: getAlerts(state),
  clusters: getClusterList(state),
});

const mapDispatchToProps = {
  fetchClustersList,
};

export default connect(mapStateToProps, mapDispatchToProps)(App);
