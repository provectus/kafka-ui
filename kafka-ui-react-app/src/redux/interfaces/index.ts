import rootReducer from 'redux/reducers';
import { store } from 'redux/store';

export * from './topic';
export * from './cluster';
export * from './consumerGroup';
export * from './schema';
export * from './loader';

export type RootState = ReturnType<typeof rootReducer>;
export type AppDispatch = typeof store.dispatch;
