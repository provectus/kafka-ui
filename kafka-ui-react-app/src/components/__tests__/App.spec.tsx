import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import { StaticRouter } from 'react-router-dom';
import { Alert } from 'redux/interfaces';
import configureStore from 'redux/store/configureStore';
import App, { AppProps } from 'components/App';
import AppContainer from 'components/AppContainer';

const fetchClustersList = jest.fn();
const store = configureStore();

describe('App', () => {
  describe('container', () => {
    it('renders view', () => {
      const wrapper = mount(
        <Provider store={store}>
          <StaticRouter>
            <AppContainer />
          </StaticRouter>
        </Provider>
      );
      expect(wrapper.exists('App')).toBeTruthy();
    });
  });
  describe('view', () => {
    const setupComponent = (props: Partial<AppProps> = {}) => (
      <Provider store={store}>
        <StaticRouter>
          <App
            isClusterListFetched
            alerts={[]}
            clusters={[]}
            fetchClustersList={fetchClustersList}
            {...props}
          />
        </StaticRouter>
      </Provider>
    );

    it('handles fetchClustersList', () => {
      const wrapper = mount(setupComponent());
      expect(wrapper.exists()).toBeTruthy();
      expect(fetchClustersList).toHaveBeenCalledTimes(1);
    });

    it('shows PageLoader until cluster list is fetched', () => {
      let component = mount(setupComponent({ isClusterListFetched: false }));
      expect(component.exists('.Layout__container PageLoader')).toBeTruthy();
      expect(component.exists('.Layout__container Switch')).toBeFalsy();

      component = mount(setupComponent({ isClusterListFetched: true }));
      expect(component.exists('.Layout__container PageLoader')).toBeFalsy();
      expect(component.exists('.Layout__container Switch')).toBeTruthy();
    });

    it('correctly renders alerts', () => {
      const alert: Alert = {
        id: 'alert-id',
        type: 'success',
        title: 'My Custom Title',
        message: 'My Custom Message',
        createdAt: 1234567890,
      };
      let wrapper = mount(setupComponent());
      expect(wrapper.exists('.Layout__alerts')).toBeTruthy();
      expect(wrapper.exists('Alert')).toBeFalsy();

      wrapper = mount(setupComponent({ alerts: [alert] }));
      expect(wrapper.exists('Alert')).toBeTruthy();
      expect(wrapper.find('Alert').length).toEqual(1);
    });

    it('matches snapshot', () => {
      const wrapper = mount(setupComponent());
      expect(wrapper).toMatchSnapshot();
    });
  });
});
