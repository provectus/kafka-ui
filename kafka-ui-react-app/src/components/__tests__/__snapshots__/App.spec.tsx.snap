// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App view matches snapshot 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <StaticRouter>
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
      staticContext={Object {}}
    >
      <App
        alerts={Array []}
        clusters={Array []}
        fetchClustersList={
          [MockFunction] {
            "calls": Array [
              Array [],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": undefined,
              },
            ],
          }
        }
        isClusterListFetched={true}
      >
        <div
          className="Layout"
        >
          <nav
            aria-label="main navigation"
            className="navbar is-fixed-top is-white Layout__header"
            role="navigation"
          >
            <div
              className="navbar-brand"
            >
              <div
                className="navbar-burger ml-0"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="button"
                tabIndex={0}
              >
                <span />
                <span />
                <span />
              </div>
              <a
                className="navbar-item title is-5 is-marginless"
                href="/ui"
              >
                Kafka UI
              </a>
              <div
                className="navbar-item"
              >
                <Version />
              </div>
            </div>
          </nav>
          <main
            className="Layout__container"
          >
            <div
              className="Layout__sidebar has-shadow has-background-white"
            >
              <Nav
                clusters={Array []}
                isClusterListFetched={true}
              >
                <aside
                  className="menu has-shadow has-background-white"
                >
                  <p
                    className="menu-label"
                  >
                    General
                  </p>
                  <ul
                    className="menu-list"
                  >
                    <li>
                      <NavLink
                        activeClassName="is-active"
                        exact={true}
                        title="Dashboard"
                        to="/ui"
                      >
                        <Link
                          aria-current={null}
                          title="Dashboard"
                          to={
                            Object {
                              "hash": "",
                              "pathname": "/ui",
                              "search": "",
                              "state": null,
                            }
                          }
                        >
                          <LinkAnchor
                            aria-current={null}
                            href="/ui"
                            navigate={[Function]}
                            title="Dashboard"
                          >
                            <a
                              aria-current={null}
                              href="/ui"
                              onClick={[Function]}
                              title="Dashboard"
                            >
                              Dashboard
                            </a>
                          </LinkAnchor>
                        </Link>
                      </NavLink>
                    </li>
                  </ul>
                  <p
                    className="menu-label"
                  >
                    Clusters
                  </p>
                </aside>
              </Nav>
            </div>
            <div
              aria-hidden="true"
              className="Layout__sidebarOverlay is-overlay"
              onClick={[Function]}
              onKeyDown={[Function]}
              tabIndex={-1}
            />
            <Switch>
              <Route
                component={[Function]}
                computedMatch={
                  Object {
                    "isExact": true,
                    "params": Object {},
                    "path": "/",
                    "url": "/",
                  }
                }
                exact={true}
                location={
                  Object {
                    "hash": "",
                    "pathname": "/",
                    "search": "",
                    "state": undefined,
                  }
                }
                path={
                  Array [
                    "/",
                    "/ui",
                    "/ui/clusters",
                  ]
                }
              >
                <Dashboard
                  history={
                    Object {
                      "action": "POP",
                      "block": [Function],
                      "createHref": [Function],
                      "go": [Function],
                      "goBack": [Function],
                      "goForward": [Function],
                      "listen": [Function],
                      "location": Object {
                        "hash": "",
                        "pathname": "/",
                        "search": "",
                        "state": undefined,
                      },
                      "push": [Function],
                      "replace": [Function],
                    }
                  }
                  location={
                    Object {
                      "hash": "",
                      "pathname": "/",
                      "search": "",
                      "state": undefined,
                    }
                  }
                  match={
                    Object {
                      "isExact": true,
                      "params": Object {},
                      "path": "/",
                      "url": "/",
                    }
                  }
                  staticContext={Object {}}
                >
                  <div
                    className="section"
                  >
                    <div
                      className="level"
                    >
                      <div
                        className="level-item level-left"
                      >
                        <Breadcrumb>
                          <nav
                            aria-label="breadcrumbs"
                            className="breadcrumb"
                          >
                            <ul>
                              <li
                                className="is-active"
                              >
                                <span
                                  className=""
                                >
                                  Dashboard
                                </span>
                              </li>
                            </ul>
                          </nav>
                        </Breadcrumb>
                      </div>
                    </div>
                    <Connect(ClustersWidget)>
                      <ClustersWidget
                        clusters={Array []}
                        dispatch={[Function]}
                        offlineClusters={Array []}
                        onlineClusters={Array []}
                      >
                        <div>
                          <h5
                            className="title is-5"
                          >
                            Clusters
                          </h5>
                          <MetricsWrapper>
                            <div
                              className="box"
                            >
                              <div
                                className="level"
                              >
                                <Indicator
                                  label="Online Clusters"
                                >
                                  <div
                                    className="level-item"
                                  >
                                    <div
                                      title="Online Clusters"
                                    >
                                      <p
                                        className="heading"
                                      >
                                        Online Clusters
                                      </p>
                                      <p
                                        className="title has-text-centered"
                                      >
                                        <span
                                          className="tag is-primary"
                                        >
                                          0
                                        </span>
                                      </p>
                                    </div>
                                  </div>
                                </Indicator>
                                <Indicator
                                  label="Offline Clusters"
                                >
                                  <div
                                    className="level-item"
                                  >
                                    <div
                                      title="Offline Clusters"
                                    >
                                      <p
                                        className="heading"
                                      >
                                        Offline Clusters
                                      </p>
                                      <p
                                        className="title has-text-centered"
                                      >
                                        <span
                                          className="tag is-danger"
                                        >
                                          0
                                        </span>
                                      </p>
                                    </div>
                                  </div>
                                </Indicator>
                                <Indicator
                                  label="Hide online clusters"
                                >
                                  <div
                                    className="level-item"
                                  >
                                    <div
                                      title="Hide online clusters"
                                    >
                                      <p
                                        className="heading"
                                      >
                                        Hide online clusters
                                      </p>
                                      <p
                                        className="title has-text-centered"
                                      >
                                        <input
                                          checked={false}
                                          className="switch is-rounded"
                                          id="switchRoundedDefault"
                                          name="switchRoundedDefault"
                                          onChange={[Function]}
                                          type="checkbox"
                                        />
                                        <label
                                          htmlFor="switchRoundedDefault"
                                        />
                                      </p>
                                    </div>
                                  </div>
                                </Indicator>
                              </div>
                            </div>
                          </MetricsWrapper>
                        </div>
                      </ClustersWidget>
                    </Connect(ClustersWidget)>
                  </div>
                </Dashboard>
              </Route>
            </Switch>
          </main>
          <div
            className="Layout__alerts"
          />
        </div>
      </App>
    </Router>
  </StaticRouter>
</Provider>
`;
